[buildkite/vllm-integration-tests](https://buildkite.com/lmcache/vllm-integration-tests/builds/6239)

Why it failed caused by without `libxcb`, `ImportError: libxcb.so.1: cannot open shared object file: No such file or directory`

```

2026-01-19 03:31:15 UTC | ===== Testing LMCache with dummy.yaml =====
-- | --
2026-01-19 03:31:15 UTC | ✅ Selected GPU #0 (CUDA_VISIBLE_DEVICES=0)
2026-01-19 03:31:15 UTC | Curl /v1/models endpoint
2026-01-19 03:35:15 UTC | OpenAI API server did not start
2026-01-19 03:35:15 UTC | (APIServer pid=1) INFO 01-19 03:31:34 [api_server.py:872] vLLM API server version 0.14.0rc2.dev137+g9a1f16da1
2026-01-19 03:35:15 UTC | (APIServer pid=1) INFO 01-19 03:31:34 [utils.py:267] non-default args: {'model_tag': 'meta-llama/Llama-3.2-1B-Instruct', 'model': 'meta-llama/Llama-3.2-1B-Instruct', 'max_model_len': 1024, 'enforce_eager': True, 'load_format': 'dummy', 'gpu_memory_utilization': 0.35, 'kv_transfer_config': KVTransferConfig(kv_connector='LMCacheConnectorV1', engine_id='2e7fceae-a77c-424b-b4af-75dfb9eb91bb', kv_buffer_device='cuda', kv_buffer_size=1000000000.0, kv_role='kv_both', kv_rank=None, kv_parallel_size=1, kv_ip='127.0.0.1', kv_port=14579, kv_connector_extra_config={}, kv_connector_module_path=None, enable_permute_local_kv=False, kv_load_failure_policy='recompute')}
2026-01-19 03:35:15 UTC | (APIServer pid=1) INFO 01-19 03:31:44 [model.py:530] Resolved architecture: LlamaForCausalLM
2026-01-19 03:35:15 UTC | (APIServer pid=1) INFO 01-19 03:31:44 [model.py:1547] Using max model len 1024
2026-01-19 03:35:15 UTC | (APIServer pid=1) INFO 01-19 03:31:44 [scheduler.py:229] Chunked prefill is enabled with max_num_batched_tokens=2048.
2026-01-19 03:35:15 UTC | (APIServer pid=1) INFO 01-19 03:31:44 [vllm.py:618] Asynchronous scheduling is enabled.
2026-01-19 03:35:15 UTC | (APIServer pid=1) INFO 01-19 03:31:44 [vllm.py:625] Disabling NCCL for DP synchronization when using async scheduling.
2026-01-19 03:35:15 UTC | (APIServer pid=1) WARNING 01-19 03:31:44 [vllm.py:653] Enforce eager set, overriding optimization level to -O0
2026-01-19 03:35:15 UTC | (APIServer pid=1) INFO 01-19 03:31:44 [vllm.py:753] Cudagraph is disabled under eager mode
2026-01-19 03:35:15 UTC | (APIServer pid=1) WARNING 01-19 03:31:44 [vllm.py:971] Turning off hybrid kv cache manager because `--kv-transfer-config` is set. This will reduce the performance of vLLM on LLMs with sliding window attention or Mamba attention. If you are a developer of kv connector, please consider supporting hybrid kv cache manager for your connector by making sure your connector is a subclass of `SupportsHMA` defined in kv_connector/v1/base.py and use --no-disable-hybrid-kv-cache-manager to start vLLM.
2026-01-19 03:35:15 UTC | Traceback (most recent call last):
2026-01-19 03:35:15 UTC | File "<string>", line 1, in <module>
2026-01-19 03:35:15 UTC | File "/usr/lib/python3.12/multiprocessing/spawn.py", line 122, in spawn_main
2026-01-19 03:35:15 UTC | exitcode = _main(fd, parent_sentinel)
2026-01-19 03:35:15 UTC | ^^^^^^^^^^^^^^^^^^^^^^^^^^
2026-01-19 03:35:15 UTC | File "/usr/lib/python3.12/multiprocessing/spawn.py", line 131, in _main
2026-01-19 03:35:15 UTC | prepare(preparation_data)
2026-01-19 03:35:15 UTC | File "/usr/lib/python3.12/multiprocessing/spawn.py", line 246, in prepare
2026-01-19 03:35:15 UTC | _fixup_main_from_path(data['init_main_from_path'])
2026-01-19 03:35:15 UTC | File "/usr/lib/python3.12/multiprocessing/spawn.py", line 297, in _fixup_main_from_path
2026-01-19 03:35:15 UTC | main_content = runpy.run_path(main_path,
2026-01-19 03:35:15 UTC | ^^^^^^^^^^^^^^^^^^^^^^^^^
2026-01-19 03:35:15 UTC | File "<frozen runpy>", line 280, in run_path
2026-01-19 03:35:15 UTC | File "/usr/lib/python3.12/pkgutil.py", line 170, in <module>
2026-01-19 03:35:15 UTC | iter_importer_modules.register(
2026-01-19 03:35:15 UTC | File "/usr/lib/python3.12/functools.py", line 892, in register
2026-01-19 03:35:15 UTC | if _is_union_type(cls):
2026-01-19 03:35:15 UTC | ^^^^^^^^^^^^^^^^^^^
2026-01-19 03:35:15 UTC | File "/usr/lib/python3.12/functools.py", line 842, in _is_union_type
2026-01-19 03:35:15 UTC | from typing import get_origin, Union
2026-01-19 03:35:15 UTC | File "/opt/venv/lib/python3.12/site-packages/cv2/typing/__init__.py", line 62, in <module>
2026-01-19 03:35:15 UTC | import cv2.mat_wrapper
2026-01-19 03:35:15 UTC | ImportError: libxcb.so.1: cannot open shared object file: No such file or directory
2026-01-19 03:35:15 UTC | (APIServer pid=1) Traceback (most recent call last):
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "/opt/venv/bin/vllm", line 10, in <module>
2026-01-19 03:35:15 UTC | (APIServer pid=1)     sys.exit(main())
2026-01-19 03:35:15 UTC | (APIServer pid=1)              ^^^^^^
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "/opt/venv/lib/python3.12/site-packages/vllm/entrypoints/cli/main.py", line 73, in main
2026-01-19 03:35:15 UTC | (APIServer pid=1)     args.dispatch_function(args)
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "/opt/venv/lib/python3.12/site-packages/vllm/entrypoints/cli/serve.py", line 60, in cmd
2026-01-19 03:35:15 UTC | (APIServer pid=1)     uvloop.run(run_server(args))
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "/opt/venv/lib/python3.12/site-packages/uvloop/__init__.py", line 96, in run
2026-01-19 03:35:15 UTC | (APIServer pid=1)     return __asyncio.run(
2026-01-19 03:35:15 UTC | (APIServer pid=1)            ^^^^^^^^^^^^^^
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "/usr/lib/python3.12/asyncio/runners.py", line 194, in run
2026-01-19 03:35:15 UTC | (APIServer pid=1)     return runner.run(main)
2026-01-19 03:35:15 UTC | (APIServer pid=1)            ^^^^^^^^^^^^^^^^
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "/usr/lib/python3.12/asyncio/runners.py", line 118, in run
2026-01-19 03:35:15 UTC | (APIServer pid=1)     return self._loop.run_until_complete(task)
2026-01-19 03:35:15 UTC | (APIServer pid=1)            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "uvloop/loop.pyx", line 1518, in uvloop.loop.Loop.run_until_complete
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "/opt/venv/lib/python3.12/site-packages/uvloop/__init__.py", line 48, in wrapper
2026-01-19 03:35:15 UTC | (APIServer pid=1)     return await main
2026-01-19 03:35:15 UTC | (APIServer pid=1)            ^^^^^^^^^^
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "/opt/venv/lib/python3.12/site-packages/vllm/entrypoints/openai/api_server.py", line 919, in run_server
2026-01-19 03:35:15 UTC | (APIServer pid=1)     await run_server_worker(listen_address, sock, args, **uvicorn_kwargs)
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "/opt/venv/lib/python3.12/site-packages/vllm/entrypoints/openai/api_server.py", line 938, in run_server_worker
2026-01-19 03:35:15 UTC | (APIServer pid=1)     async with build_async_engine_client(
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "/usr/lib/python3.12/contextlib.py", line 210, in __aenter__
2026-01-19 03:35:15 UTC | (APIServer pid=1)     return await anext(self.gen)
2026-01-19 03:35:15 UTC | (APIServer pid=1)            ^^^^^^^^^^^^^^^^^^^^^
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "/opt/venv/lib/python3.12/site-packages/vllm/entrypoints/openai/api_server.py", line 146, in build_async_engine_client
2026-01-19 03:35:15 UTC | (APIServer pid=1)     async with build_async_engine_client_from_engine_args(
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "/usr/lib/python3.12/contextlib.py", line 210, in __aenter__
2026-01-19 03:35:15 UTC | (APIServer pid=1)     return await anext(self.gen)
2026-01-19 03:35:15 UTC | (APIServer pid=1)            ^^^^^^^^^^^^^^^^^^^^^
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "/opt/venv/lib/python3.12/site-packages/vllm/entrypoints/openai/api_server.py", line 187, in build_async_engine_client_from_engine_args
2026-01-19 03:35:15 UTC | (APIServer pid=1)     async_llm = AsyncLLM.from_vllm_config(
2026-01-19 03:35:15 UTC | (APIServer pid=1)                 ^^^^^^^^^^^^^^^^^^^^^^^^^^
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "/opt/venv/lib/python3.12/site-packages/vllm/v1/engine/async_llm.py", line 202, in from_vllm_config
2026-01-19 03:35:15 UTC | (APIServer pid=1)     return cls(
2026-01-19 03:35:15 UTC | (APIServer pid=1)            ^^^^
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "/opt/venv/lib/python3.12/site-packages/vllm/v1/engine/async_llm.py", line 129, in __init__
2026-01-19 03:35:15 UTC | (APIServer pid=1)     self.engine_core = EngineCoreClient.make_async_mp_client(
2026-01-19 03:35:15 UTC | (APIServer pid=1)                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "/opt/venv/lib/python3.12/site-packages/vllm/v1/engine/core_client.py", line 122, in make_async_mp_client
2026-01-19 03:35:15 UTC | (APIServer pid=1)     return AsyncMPClient(*client_args)
2026-01-19 03:35:15 UTC | (APIServer pid=1)            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "/opt/venv/lib/python3.12/site-packages/vllm/v1/engine/core_client.py", line 819, in __init__
2026-01-19 03:35:15 UTC | (APIServer pid=1)     super().__init__(
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "/opt/venv/lib/python3.12/site-packages/vllm/v1/engine/core_client.py", line 479, in __init__
2026-01-19 03:35:15 UTC | (APIServer pid=1)     with launch_core_engines(vllm_config, executor_class, log_stats) as (
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "/usr/lib/python3.12/contextlib.py", line 144, in __exit__
2026-01-19 03:35:15 UTC | (APIServer pid=1)     next(self.gen)
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "/opt/venv/lib/python3.12/site-packages/vllm/v1/engine/utils.py", line 918, in launch_core_engines
2026-01-19 03:35:15 UTC | (APIServer pid=1)     wait_for_engine_startup(
2026-01-19 03:35:15 UTC | (APIServer pid=1)   File "/opt/venv/lib/python3.12/site-packages/vllm/v1/engine/utils.py", line 977, in wait_for_engine_startup
2026-01-19 03:35:15 UTC | (APIServer pid=1)     raise RuntimeError(
2026-01-19 03:35:15 UTC | (APIServer pid=1) RuntimeError: Engine core initialization failed. See root cause above. Failed core proc(s): {'EngineCore_DP0': 1}
2026-01-19 03:35:15 UTC | → Cleaning up Docker containers and ports...
2026-01-19 03:35:15 UTC | - Killing and removing container: e61883e9c16ae37579fe22ab4dc829d8efa7fb758b22a946471e750036bd1564
2026-01-19 03:35:15 UTC | Error response from daemon: Cannot kill container: e61883e9c16ae37579fe22ab4dc829d8efa7fb758b22a946471e750036bd1564: Container e61883e9c16ae37579fe22ab4dc829d8efa7fb758b22a946471e750036bd1564 is not running
2026-01-19 03:35:15 UTC | e61883e9c16ae37579fe22ab4dc829d8efa7fb758b22a946471e750036bd1564
2026-01-19 03:35:15 UTC | - Killing and removing port: 8000
2026-01-19 03:35:15 UTC | 🚨 Error: The command exited with status 1


```

--- a/lmcache/v1/lookup_client/lmcache_lookup_client.py
+++ b/lmcache/v1/lookup_client/lmcache_lookup_client.py
@@ -330,11 +330,14 @@ class LMCacheLookupServer:
                 identity = frames[0].bytes
                 # frames[1] is the empty delimiter frame from REQ socket
                 data_frames = frames[2:]
+                if len(data_frames) < 2:
+                    logger.warning("Malformed request received: not enough frames.")
+                    continue
                 lookup_id = data_frames[-2].bytes.decode("utf-8")
                 request_configs_str = data_frames[-1].bytes.decode("utf-8")
-                request_configs = None
-                if request_configs_str != "":
-                    request_configs = json.loads(request_configs_str)
+                request_configs = (
+                    json.loads(request_configs_str) if request_configs_str else None
+                )
                 if not self.enable_blending:
                     hash_frames = data_frames[0]
                     offset_frames = data_frames[1]

--- a/lmcache/v1/lookup_client/lmcache_lookup_client.py
+++ b/lmcache/v1/lookup_client/lmcache_lookup_client.py
@@ -307,7 +307,7 @@ class LMCacheLookupServer:
             self.ctx,
             socket_path,
             "ipc",
-            zmq.REP,  # type: ignore[attr-defined]
+            zmq.ROUTER,  # type: ignore[attr-defined]
             "bind",
         )
         # Set socket timeout to allow periodic check of running flag
@@ -325,14 +325,19 @@ class LMCacheLookupServer:
                 except zmq.Again:
                     # Timeout occurred, check running flag and continue
                     continue
-                lookup_id = frames[-2].bytes.decode("utf-8")
-                request_configs_str = frames[-1].bytes.decode("utf-8")
+                # ROUTER socket prepends identity frame and empty delimiter
+                # frames[0] = identity, frames[1] = empty delimiter, frames[2:] = data
+                identity = frames[0].bytes
+                # frames[1] is the empty delimiter frame from REQ socket
+                data_frames = frames[2:]
+                lookup_id = data_frames[-2].bytes.decode("utf-8")
+                request_configs_str = data_frames[-1].bytes.decode("utf-8")
                 request_configs = None
                 if request_configs_str != "":
                     request_configs = json.loads(request_configs_str)
                 if not self.enable_blending:
-                    hash_frames = frames[0]
-                    offset_frames = frames[1]
+                    hash_frames = data_frames[0]
+                    offset_frames = data_frames[1]
                     hashes = self.decoder.decode(hash_frames)
                     offsets = self.decoder.decode(offset_frames)
                     result = self.lmcache_engine.lookup(
@@ -343,7 +348,7 @@ class LMCacheLookupServer:
                         request_configs=request_configs,
                     )
                 else:
-                    token_frames = frames[0]
+                    token_frames = data_frames[0]
                     tokens = self.decoder.decode(token_frames)
                     result = self.lmcache_engine.lookup(
                         tokens=tokens,
@@ -352,7 +357,8 @@ class LMCacheLookupServer:
                         request_configs=request_configs,
                     )
                 response = result.to_bytes(4, "big")
-                self.socket.send(response)
+                # ROUTER requires identity frame and empty delimiter for reply
+                self.socket.send_multipart([identity, b"", response])
 
         logger.info("lmcache lookup server start on %s", socket_path)
         self.thread = threading.Thread(target=process_request, daemon=True)

